### PR TITLE
A0-4560: Changes in dag reconstruction to include parent round

### PR DESCRIPTION
**Pull request by Marcin-Radecki**
This PR is a follow up after https://github.com/Cardinal-Cryptography/AlephBFT/commit/71f03144faa5856c85f477142c013fe36659b9e3, that introduced notion of parent round in the Extender. 

Here we modify `ControlHash`  accordingly. Previously, it was using `NodeSubset` which was just a `Vec<bool>` so a true/false whether given node index created a parent unit. Since we need now also the parent round, all code that reconstructs unit now have notion of a parent round. 

`NodeSubset` is not removed as it is still used in the codebase.

This change is actually a no-op to the algo - you can see in the PR removed code `round - 1` to just parent round from unit store, however in practice it is still `round - 1` as the changes to Creator are not done yet. This will be done in the next PR.

### CODE REVIEW

File: consensus/src/units/mod.rs [hunk 1]
```diff
@@ -56,32 +56,35 @@ impl Display for UnitCoord {
...
+    /// Calculate parent control hash, which includes all parent hashes into account.
+    pub(crate) fn combine_hashes(parent_map: &NodeMap<(H::Hash, Round)>) -> H::Hash {
+        // we include parent rounds with calculating hash but this is okay - we cannot
+        // have two units with the same hash but different rounds
         parent_map.using_encoded(H::hash)
     }
 
-    pub(crate) fn parents(&self) -> impl Iterator<Item = NodeIndex> + '_ {
-        self.parents_mask.elements()
+    pub(crate) fn parents(&self) -> impl Iterator<Item = (NodeIndex, &(H::Hash, Round))> + '_ {
```
[the remark this answers is not in the document]
Comment (timorleph) [reply consensus/src/units/mod.rs:78]: AAaa,a wait, this is much more wrong, why is there `H::Hash` at all? It makes everything much _much_ bigger and makes the control hash pointless – the entire idea with it was to _not_ send hashes for every unit, just a single combined one. And I _almost_ missed this, I was just about to submit a review with just the other comments. :woozy_face: 

Anyway, remove the hashes from the control hash.

[the remark this answers is not in the document]
Comment (timorleph) [reply consensus/src/units/mod.rs:78]: This actually makes the first comment in this thread a much more reasonable suggestion – cloning hashes is annoying, but `UnitCoord` is tiny (and possibly even `Copy`?).

**Review by timorleph — CHANGES_REQUESTED** [review]
Oof, see my last comments for what's really wrong with this PR. Except for that one major thing, the rest is quite reasonable, even if in places I'm a bit sad about the juggling between `NodeMap`s and `UnitCoords` (I don't think this can be made better, so no inline comments about it, unless I forgot to delete one :P).

**Review by DamianStraszak — CHANGES_REQUESTED** [review]
Initial pass, I will take another look after the hashes are removed from control hash.

File: consensus/src/units/mod.rs [hunk 2]
```diff
@@ -56,32 +56,35 @@ impl Display for UnitCoord {
 /// parents
 #[derive(Clone, Eq, PartialEq, Hash, Debug, Decode, Encode)]
 pub struct ControlHash<H: Hasher> {
-    pub(crate) parents_mask: NodeSubset,
+    pub(crate) parents_round_lookup: NodeMap<(H::Hash, Round)>,
```
[the remark this answers is not in the document]
Comment (timorleph) [reply consensus/src/units/mod.rs:59]: > It should be something like `Vec<Option<NodeIndex>>`

Good news, `NodeMap<Round>` is exactly that! (Or rather with `Round` in place of `NodeIndex`, but I'm pretty sure that's what you meant). I'm not sure we can make this significantly smaller with a special `Codec` implementation though, since now we actually need to send the rounds rather than just bits. :/

File: consensus/src/units/mod.rs [hunk 3]
```diff
@@ -56,32 +56,35 @@ impl Display for UnitCoord {
...
+            parents_round_lookup: parents_round_lookup.clone(),
+            combined_hash: Self::combine_hashes(parents_round_lookup),
         }
     }
 
-    pub(crate) fn combine_hashes(parent_map: &NodeMap<H::Hash>) -> H::Hash {
+    /// Calculate parent control hash, which includes all parent hashes into account.
+    pub(crate) fn combine_hashes(parent_map: &NodeMap<(H::Hash, Round)>) -> H::Hash {
+        // we include parent rounds with calculating hash but this is okay - we cannot
+        // have two units with the same hash but different rounds
```
[the remark this answers is not in the document]
Comment (timorleph) [reply consensus/src/units/mod.rs]: Doesn't seem like you did, did you miss a commit?

**Review by timorleph — APPROVED** [review]
Looks good imo, caught you on one lie. ;p

File: consensus/src/dag/reconstruction/mod.rs [hunk 4]
```diff
@@ -21,22 +21,11 @@ pub struct ReconstructedUnit<U: Unit> {
 
 impl<U: Unit> ReconstructedUnit<U> {
     /// Returns a reconstructed unit if the parents agree with the hash, errors out otherwise.
-    pub fn with_parents(unit: U, parents: NodeMap<HashFor<U>>) -> Result<Self, U> {
+    pub fn with_parents(unit: U, parents: NodeMap<(HashFor<U>, Round)>) -> Result<Self, U> {
```
Comment (timorleph) [inline consensus/src/dag/reconstruction/mod.rs:24]: Maybe add some tests for this function with actual nontrivial `Round` values, e.g. wrong rounds with matching hashes, everything correct, but there are some ancient units etc.

File: consensus/src/dag/reconstruction/parents.rs [hunk 5]
```diff
@@ -3,13 +3,14 @@ use crate::{
...
 };
+use aleph_bft_types::Round;
 use std::collections::{hash_map::Entry, HashMap};
 
 /// A unit in the process of reconstructing its parents.
 #[derive(Debug, PartialEq, Eq, Clone)]
 enum ReconstructingUnit<U: Unit> {
     /// We are trying to optimistically reconstruct the unit from potential parents we get.
-    Reconstructing(U, NodeMap<HashFor<U>>),
+    Reconstructing(U, NodeMap<(HashFor<U>, Round)>),
```
Comment (timorleph) [inline consensus/src/dag/reconstruction/parents.rs:13]: Some tests that check that this addition works would be nice as well.

**Review by timorleph — APPROVED** [review]
I hate that GitHub confuses reviews with comments. ;_;

File: consensus/src/units/testing.rs [hunk 6]
```diff
@@ -135,15 +135,18 @@ fn random_initial_reconstructed_units(
         .collect()
 }
 
-fn parent_map<U: Unit<Hasher = Hasher64>>(parents: &Vec<U>) -> NodeMap<Hash64> {
+fn parent_map<U: Unit<Hasher = Hasher64>>(
+    parents: &Vec<U>,
+    parent_round: Round,
```
Comment (timorleph) [inline consensus/src/units/testing.rs:140]: Wait, this makes very little sense, you can just use `parent.round()` and make this function much more generic, no?

File: consensus/src/testing/crash_recovery.rs [hunk 7]
```diff
@@ -132,10 +132,10 @@ fn verify_backup(buf: &mut &[u8]) -> HashSet<UnitCoord> {
         let unit = <UncheckedSignedUnit<Hasher64, Data, Signature>>::decode(buf).unwrap();
         let full_unit = unit.as_signable();
         let coord = full_unit.coord();
-        let parent_ids = &full_unit.as_pre_unit().control_hash().parents_mask;
+        let parent_ids = &full_unit.as_pre_unit().control_hash().parents;
 
-        for parent_id in parent_ids.elements() {
-            let parent = UnitCoord::new(coord.round() - 1, parent_id);
+        for (parent_id, &round) in parent_ids.iter() {
+            let parent = UnitCoord::new(round, parent_id);
```
Comment (timorleph) [inline consensus/src/testing/crash_recovery.rs:138]: You don't need this line if you iterate over `control_hash().parents()` rather than touching internals (probably a leftover from other changes after previous reviews).

File: consensus/src/units/validator.rs [hunk 8]
```diff
@@ -138,7 +138,7 @@ impl<K: Keychain> Validator<K> {
                     return Err(ValidationError::NotEnoughParents(pre_unit.clone()));
                 }
                 let control_hash = &pre_unit.control_hash();
-                if !control_hash.parents_mask[pre_unit.creator()] {
+                if control_hash.parents.get(pre_unit.creator()).is_none() {
```
Comment (timorleph) [inline consensus/src/units/validator.rs:141]: Oof, as @DamianStraszak pointed out through other channels this is wrong, together with a significant part of this method. So, the problems:
1. The parent of this creator should not only exist, but also be of round exactly `n-1`.
2. We should check that we have at least `threshold` parents of round exactly `n-1`.
3. We should check that there are no parents of rounds _higher_ than `n-1`.

All this due to `ControlHash` suddenly becoming much less trivial. In fact, I would recommend moving this `ControlHash` validation logic into a method of `ControlHash`, like `pub fn validate(&self, unit_round: Round) -> Result<(), Error>` or something. Perhaps even create a submodule for `ControlHash`, since it became so nontrivial, and then add some tests there.

**Review by timorleph — CHANGES_REQUESTED** [review]
Damian found some serious problems, but these seem to actually be the last one this time. <_<"

File: consensus/src/testing/crash_recovery.rs (hunk 7 above)
Comment (Marcin-Radecki) [reply consensus/src/testing/crash_recovery.rs:138]: Correct, changed

File: consensus/src/units/testing.rs (hunk 6 above)
Comment (Marcin-Radecki) [reply consensus/src/units/testing.rs:140]: You're right, I don't know why I did this in the first place

**Review by Marcin-Radecki — COMMENTED** [review]
I pushed changes for all remarks except control hash validation

File: consensus/src/units/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply consensus/src/units/mod.rs]: This should be now cleared, as this comment pointed to different commit in the PR, but for some reasons GH still shows it there

File: consensus/src/dag/reconstruction/parents.rs (hunk 5 above)
Comment (timorleph) [reply consensus/src/dag/reconstruction/parents.rs:13]: Oh, no, that's not what I meant, sorry for not being clear, adding methods just to test them doesn't seem great. <_<"

I wanted there to be a test in a similar vein to the one in `dag::reconstruction`, i.e. in this case passing a unit with wrong round in `add_unit` I guess?

Comment (Marcin-Radecki) [reply consensus/src/dag/reconstruction/parents.rs:13]: Okay, let me think about it, as when the unit is created I cannot easily change its round, can I. But you're right I should revert this `parent_round` method.

File: consensus/src/dag/reconstruction/parents.rs [hunk 9]
```diff
@@ -178,6 +180,7 @@ impl<U: Unit> Reconstruction<U> {
                             unit_hash,
                             parent_coord.creator(),
                             *parent_hash,
+                            parent_coord.round(),
```
Comment (Marcin-Radecki) [inline consensus/src/dag/reconstruction/parents.rs:183]: While writing some unit test to cover scenario in which a unit pretend to have different coord - there are two distinc sources of information passed now down to `with_parents:
1. `parent_hash` from the line above comes from `units_by_coord` map, which is filled what a unit claims to have its coord
2. `parent_coord.round` is really taken from its control hash
and since we don't have any validation between unit coord and its control hash, at least in the tests I can pass here mismatched data, ie see commit https://github.com/Cardinal-Cryptography/AlephBFT/pull/514/commits/aef31c8892f6aa4345c75729986ca367ce91afc6 . Now my question is, is this line okay, or should I change `units_by_coord` to keep both hash and round, so sources of both value is the same?

File: consensus/src/dag/reconstruction/parents.rs (hunk 5 above)
Comment (Marcin-Radecki) [reply consensus/src/dag/reconstruction/parents.rs:13]: This should be now addressed.

File: consensus/src/units/control_hash.rs [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    /// Returns number of all members in abft consensus
+    pub(crate) fn n_members(&self) -> NodeCount {
+        self.parents.size()
+    }
+
+    /// Validate
+    pub fn validate(&self, unit_coord: UnitCoord) -> Result<(), Error> {
+        assert!(unit_coord.round > 0, "Round must be greater than 0");
```
Comment (Marcin-Radecki) [inline consensus/src/units/control_hash.rs:57]: We should validate also initial units here

File: consensus/src/units/control_hash.rs [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+        assert_eq!(
+            ch.validate(UnitCoord::new(3, NodeIndex(1)))
+                .expect_err("validate() should return error, returned Ok(()) instead"),
+            Error::NotEnoughParentsForRound(2)
+        );
+    }
+
+    #[test]
+    fn given_control_hash_when_there_are_parents_from_greater_rounds_then_err_is_returned_from_validate(
+    ) {
```
Comment (Marcin-Radecki) [inline consensus/src/units/control_hash.rs:242]: Add a test discussed offline

File: consensus/src/dag/reconstruction/mod.rs [hunk 12]
```diff
@@ -308,6 +299,50 @@ mod test {
         }
     }
 
+    #[test]
+    fn given_dag_units_with_wrong_when_reconstructing_then_unit_is_not_reconstructed() {
```
Comment (Marcin-Radecki) [inline consensus/src/dag/reconstruction/mod.rs:303]: This test does not make sense since it's done after validation

File: consensus/src/units/control_hash.rs (hunk 10 above)
Comment (Marcin-Radecki) [reply consensus/src/units/control_hash.rs:127]: Done

File: consensus/src/units/control_hash.rs (hunk 11 above)
Comment (Marcin-Radecki) [reply consensus/src/units/control_hash.rs:356]: Done, but please bear with me I added them in ControlHash module only.

File: consensus/src/dag/reconstruction/mod.rs (hunk 12 above)
Comment (Marcin-Radecki) [reply consensus/src/dag/reconstruction/mod.rs]: Removed

File: consensus/src/dag/reconstruction/parents.rs (hunk 9 above)
Comment (Marcin-Radecki) [reply consensus/src/dag/reconstruction/parents.rs:183]: This comment is not valid anymore, since I did not take into account validation, which filters this scenario

File: consensus/src/units/validator.rs (hunk 8 above)
Comment (Marcin-Radecki) [reply consensus/src/units/validator.rs]: This should be now addressed.

File: consensus/src/units/control_hash.rs [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+            Error::NotDescendantOfPreviousUnit(node_index) => {
+                write!(
+                    f,
+                    "non-zero round unit is not descendant of its creator's previous unit, creator index: {:?}",
+                    node_index)
+            }
+            Error::DescendantOfPreviousUnitButWrongRound(round) => {
+                write!(
+                f,
+                "non-zero round unit is descendant of its creator's previous unit, but has wrong round {:?}",
```
Comment (timorleph) [inline consensus/src/units/control_hash.rs:42]: Nit: technically "previous" means "exactly preceeding", so wrong round in this case means it's _not_ a descendant of the previous unit.

File: consensus/src/units/control_hash.rs [hunk 14]
```diff
@@ -0,0 +1,412 @@
...
+        }
+    }
+}
+
+/// Combined hashes of the parents of a unit together with the set of indices of creators of the
+/// parents. By parent here we mean a parent hash and its round.
+#[derive(Clone, Eq, PartialEq, Hash, Debug, Decode, Encode)]
+pub struct ControlHash<H: Hasher> {
+    pub(crate) parents: NodeMap<Round>,
+    pub(crate) combined_hash: H::Hash,
```
Comment (timorleph) [inline consensus/src/units/control_hash.rs:69]: Do these really have to be `pub`? You seem to have proper access and creation methods for everything, no?

File: consensus/src/units/control_hash.rs [hunk 15]
```diff
@@ -0,0 +1,412 @@
...
+}
+
+impl<H: Hasher> ControlHash<H> {
+    /// Creates new control hash from parents hashes and rounds
+    pub(crate) fn new(parents_with_rounds_and_hashes: &NodeMap<(H::Hash, Round)>) -> Self {
```
Comment (timorleph) [inline consensus/src/units/control_hash.rs:74]: Speaking of `pub` – I'm pretty sure you can just remove the `(crate)` everywhere here, it's ugly and serves no purpose.

File: consensus/src/units/control_hash.rs [hunk 16]
```diff
@@ -0,0 +1,412 @@
...
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use crate::units::control_hash::Error;
+    use crate::units::UnitCoord;
+    use crate::{units::ControlHash, NodeCount, NodeIndex};
```
Comment (timorleph) [inline consensus/src/units/control_hash.rs:178]: Combine these (I thought clippy or fmt complained about this, but apparently not?).

File: consensus/src/units/control_hash.rs [hunk 17]
```diff
@@ -0,0 +1,412 @@
...
+    #[test]
+    fn given_control_hash_then_basic_properties_are_correct() {
+        let parent_map = vec![
+            Some(([0; 8], 2)),
+            None,
+            Some(([2; 8], 2)),
+            Some(([3; 8], 2)),
+            Some(([4; 8], 2)),
+            Some(([5; 8], 2)),
+            Some(([5; 8], 1)),
```
Comment (timorleph) [inline consensus/src/units/control_hash.rs:202]: Uhm, please don't pass multiple different units with the same hash – this is not a situation we want to test, at least not if we are not testing some weird corner case on purpose. Some tests below also have this problem.

(It currently works, but we might at some point assume that this is not the case, which could break these tests for no good reason.)

File: consensus/src/units/control_hash.rs [hunk 18]
```diff
@@ -0,0 +1,412 @@
...
+            ch.validate(UnitCoord::new(0, NodeIndex(1)))
+                .expect_err("validate() should return error, returned Ok(()) instead"),
+            Error::RoundZeroWithSomeParents(NodeCount(parent_map.item_count()))
+        );
+    }
+
+    #[test]
+    fn given_initial_round_when_validate_with_different_node_count_then_validate_returns_err() {
+        // Correct control hash for initial unit with count = 4 is
+        // [16, 0, 0, 0, 0, 129, 99, 217, 65, 183, 158, 24, 201];
```
Comment (timorleph) [inline consensus/src/units/control_hash.rs:249]: Hmm, perhaps it would be better to construct the control hash, encode it, and then flip a bit in the latter part of it, rather than manually putting in the encoding? It will still require a comment, to make clear that you are flipping a bit in the control hash, but it will at least be less fragile to potential future changes to `ControlHash`.

File: consensus/src/units/control_hash.rs [hunk 19]
```diff
@@ -0,0 +1,412 @@
...
+    fn given_initial_round_when_validate_called_for_initial_unit_then_function_panics() {
+        let parent_map = vec![
+            Some(([0; 8], 2)),
+            None,
+            Some(([2; 8], 2)),
+            Some(([3; 8], 2)),
+        ]
+        .into();
+        let ch = ControlHash::<Hasher64>::new(&parent_map);
+        let _ = ch.validate_non_initial_round(UnitCoord::new(0, NodeIndex(1)));
```
Comment (timorleph) [inline consensus/src/units/control_hash.rs:298]: No need to test an internal function like that imo.

File: consensus/src/units/control_hash.rs (hunk 14 above)
Comment (timorleph) [reply consensus/src/units/control_hash.rs:69]: Oh, not for `combined_hash`, but it would probably be nice if there was one.

File: consensus/src/units/control_hash.rs [hunk 20]
```diff
@@ -0,0 +1,412 @@
...
+        let borked_hash_of_fourth_round_unit =
+            ControlHash::<Hasher64>::new(&parents_from_round_three_but_one_hash_replaced);
+        assert_ne!(
+            borked_hash_of_fourth_round_unit,
+            control_hash_of_fourth_round_unit
+        );
+
+        let mut parents_from_round_three_but_one_unit_replaced = parents_from_round_three.clone();
+        parents_from_round_three_but_one_unit_replaced
+            .insert(NodeIndex(2), all_parents_from_round_three[3].unwrap());
```
Comment (timorleph) [inline consensus/src/units/control_hash.rs:400]: I think this was supposed to change the `round`, not replace the whole thing.

File: consensus/src/units/validator.rs [hunk 21]
```diff
@@ -34,20 +31,10 @@ impl<H: Hasher, D: Data, S: Signature> Display for ValidationError<H, D, S> {
...
-                "nonzero round unit with only {:?} parents: {:?}",
-                pu.n_parents(),
-                pu
-            ),
-            NotDescendantOfPreviousUnit(pu) => write!(
+            ParentValidationFailed(pu, control_hash_error) => write!(
                 f,
-                "nonzero round unit is not descendant of its creator's previous unit: {:?}",
-                pu
+                "parent validation failed for nit: {:?}. Internal error: {:?}",
```
Comment (timorleph) [inline consensus/src/units/validator.rs:36]: Nit: s/nit/unit :>

Also just "...for unit {:?}: {}" will probably look nicer. (Note the `{}` for the error.)

File: consensus/src/units/validator.rs [hunk 22]
```diff
@@ -118,40 +101,18 @@ impl<K: Keychain> Validator<K> {
...
-                if !control_hash.parents_mask[pre_unit.creator()] {
-                    return Err(ValidationError::NotDescendantOfPreviousUnit(
-                        pre_unit.clone(),
-                    ));
-                }
-            }
-        }
+        let unit_coord = UnitCoord::new(pre_unit.round(), pre_unit.creator());
+        pre_unit.control_hash.validate(unit_coord).map_err(|e| {
+            ValidationError::<H, D, K::Signature>::ParentValidationFailed(pre_unit.clone(), e)
```
Comment (timorleph) [inline consensus/src/units/validator.rs:106]: I don't think you need the turbofish here.

File: consensus/src/units/validator.rs [hunk 23]
```diff
@@ -118,40 +101,18 @@ impl<K: Keychain> Validator<K> {
...
+        })?;
         Ok(su)
     }
 }
 
 #[cfg(test)]
 mod tests {
     use super::{ValidationError::*, Validator as GenericValidator};
+    use crate::units::ControlHashError;
```
Comment (timorleph) [inline consensus/src/units/validator.rs:115]: Join the import with the one below.

File: consensus/src/units/control_hash.rs [hunk 24]
```diff
@@ -0,0 +1,412 @@
...
+        self.previous_round_have_enough_parents(unit_coord.round)?;
+        self.check_if_parents_greater_than_previous_round(unit_coord.round)?;
+
+        Ok(())
+    }
+
+    fn check_if_parents_greater_than_previous_round(&self, round: Round) -> Result<(), Error<H>> {
+        let parents_greater_than_previous_round = self
+            .parents()
+            .filter(|&parent| parent.round > round - 1)
```
Comment (timorleph) [inline consensus/src/units/control_hash.rs:139]: You can just use `any` here.

File: consensus/src/units/control_hash.rs [hunk 25]
```diff
@@ -0,0 +1,412 @@
+use crate::{units::UnitCoord, Hasher, NodeCount, NodeIndex, NodeMap, Round};
+use codec::{Decode, Encode};
+use std::fmt::{Display, Formatter, Result as FmtResult};
+use std::hash::Hash;
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply consensus/src/units/control_hash.rs]: Done

File: consensus/src/units/control_hash.rs (hunk 13 above)
Comment (Marcin-Radecki) [reply consensus/src/units/control_hash.rs]: Changed

File: consensus/src/units/control_hash.rs [hunk 26]
```diff
@@ -0,0 +1,412 @@
...
+                write!(
+                    f,
+                    "non-zero round unit has not enough parents from the round {:?}",
+                    round
+                )
+            }
+            Error::ParentsHigherThanRound(round) => {
+                write!(
+                    f,
+                    "non-zero round unit has parents higher than round {:?}",
```
[the remark this answers is not in the document]
Comment (Marcin-Radecki) [reply consensus/src/units/control_hash.rs]: Yes, removed those prefixes

File: consensus/src/units/control_hash.rs (hunk 14 above)
Comment (Marcin-Radecki) [reply consensus/src/units/control_hash.rs]: Removed `pub(create)`, added getter `control_hash()` and fixed some UT which directly used `control_hash` field

File: consensus/src/units/control_hash.rs (hunk 15 above)
Comment (Marcin-Radecki) [reply consensus/src/units/control_hash.rs]: Good point, removed

File: consensus/src/units/control_hash.rs (hunk 24 above)
Comment (Marcin-Radecki) [reply consensus/src/units/control_hash.rs]: Good point, done

File: consensus/src/units/control_hash.rs (hunk 16 above)
Comment (Marcin-Radecki) [reply consensus/src/units/control_hash.rs]: Done

File: consensus/src/units/control_hash.rs (hunk 17 above)
Comment (Marcin-Radecki) [reply consensus/src/units/control_hash.rs]: This was a mistake propagated by a copy-paste - now fixed.

File: consensus/src/units/control_hash.rs (hunk 18 above)
Comment (Marcin-Radecki) [reply consensus/src/units/control_hash.rs:241]: `I'm now starting from encoded control hash, though at some point I need to assumed some internals of encoding, hopefully though this version is better.

File: consensus/src/units/control_hash.rs (hunk 19 above)
Comment (Marcin-Radecki) [reply consensus/src/units/control_hash.rs]: Good point, test removed

File: consensus/src/units/control_hash.rs (hunk 20 above)
Comment (Marcin-Radecki) [reply consensus/src/units/control_hash.rs]: Right, changed

File: consensus/src/units/validator.rs (hunk 21 above)
Comment (Marcin-Radecki) [reply consensus/src/units/validator.rs]: Done

File: consensus/src/units/validator.rs (hunk 22 above)
Comment (Marcin-Radecki) [reply consensus/src/units/validator.rs]: Right, removed

File: consensus/src/units/validator.rs (hunk 23 above)
Comment (Marcin-Radecki) [reply consensus/src/units/validator.rs]: Done

**Review by DamianStraszak — APPROVED** [review]
